Add support for Rule Based Promotions for Choice of Bonus Products

sf-shikhar-prasoon opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->

# Description
Add support for Rule Base Promotions to Choice of Bonus Products.

Last release we added support for Choice of Bonus Products where the shopper is presented a set of choices to select bonus products from. 

Currently only list based promotions are supported. After merging this PR, rule based promotions will be supported.

Currently, merchants have to explicitly list the products that are offered as choices. After merging this PR, merchants can configure rules that will search for all products that satisfy the promotion rule and create the list to be presented to the shopper.


https://github.com/user-attachments/assets/07213ce3-68a5-408f-b609-72b6d637d6c2



<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [x] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

### Features:

- New Hook: useRuleBasedBonusProducts - Fetches bonus products using ShopperSearch productSearch endpoint with promotionId
- Detection Logic: isRuleBasedPromotion() - Identifies rule-based vs list-based promotions (empty bonusProducts array indicator)
- Enhanced Discovery: Updated discovery.js to handle both promotion types via ruleBasedProductsMap parameter
- Business Logic: New utilities in business-logic.js for rule-based promotion detection

### Affected Components:

- app/hooks/use-rule-based-bonus-products.js (NEW - 50 lines)
- app/utils/bonus-product/discovery.js (+81 lines)
- app/utils/bonus-product/business-logic.js (+42 lines)
- app/hooks/use-derived-product.js (updated integration)
- 

# How to Test-Drive This PR

- [MRT Env](https://258demoprj-cc-sharks.sfdc-3vx9f4-commerceecom.exp-delivery-staging.com/) . Qualifying product: Search for "25493613M"
- To test locally, run the app locally
- Go to product [25493613M](http://localhost:3000/global/en-GB/product/25493613M?color=JJS91A5&size=004&pid=701642842668M) Long Center Seam Skirt
- Add it to cart and click on "Select Bonus Products"
- Verify that you see a list of bonus products that matches the promotion (shown below)
- You can also setup more rule based promotions and test them
The promotion:
<img width="1005" height="496" alt="Screenshot 2025-10-23 at 10 13 56 AM" src="https://github.com/user-attachments/assets/6817e947-308c-4170-91b0-8a6a0164626b" />

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [x] Changes are covered by test cases
- [x] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

sf-shikhar-prasoon committed:
Merge pull request #3392 from SalesforceCommerceCloud/develop

update feature branch `bonus-products-rule-based-v3` with develop branch

sf-shikhar-prasoon committed:
Merge pull request #3395 from SalesforceCommerceCloud/develop

update feature branch `bonus-products-rule-based-v3` with develop branch

sf-shikhar-prasoon committed:
Merge pull request #3406 from SalesforceCommerceCloud/develop

update feature branch `bonus-products-rule-based-v3` with develop branch

sf-shikhar-prasoon committed:
@W-18957985- Product Search Integration & Rule-Based Detection (#3396)

* 1. use search endpoint

* 2 add rule based detection

* 3. update discovery logic

* add tests, update bundle size limit, update changelog

sf-shikhar-prasoon committed:
@W-19980361 Rule based bonus products- Work Item 2- update display logic (#3408)

* 1. use search endpoint

* 2 add rule based detection

* 3. update discovery logic

* cleanup. remove unused code

* add tests, update bundle size limit, update changelog

cc-prodsec commented:
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-critical.svg) Critical | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg) High | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-medium.svg) Medium | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-low.svg) Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/eb2604ff-8f91-41ea-b7a1-898443e8aa54/license?refs=d0fa4fe33bc5707767770c3162d7757eb4999e8c&source=prChecksComment) |
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/eb2604ff-8f91-41ea-b7a1-898443e8aa54?refs=d0fa4fe33bc5707767770c3162d7757eb4999e8c&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

sf-shikhar-prasoon committed:
Merge pull request #3419 from SalesforceCommerceCloud/develop

Update feature branch with latest from develop branch

patricksullivansf commented on packages/template-retail-react-app/CHANGELOG.md:
```diff
@@ -1,3 +1,6 @@
+## v8.3.0
+- Add support for Rule Based Promotions for Choice of Bonus Products.
+
 ## v8.2.0-dev (Sep 26, 2025)
```
Seems this feature should be targeting v8.2.0-dev or v8.3.0-dev but not 8.3.0?

sf-shikhar-prasoon replied:
Yes, I've asked @shethj what the version should be. I just put something since I was not sure. I'll update it as soon as I get a confirmation

shethj replied:
Next release version will be 8.2.0 so all changelog should be under 8.2.0-dev
The changelog versions will update during the release process.

shethj replied:
Also, please follow the changelog format and add in a link to the PR that is related to the changes.
Thanks :D

sf-shikhar-prasoon committed:
Merge branch 'develop' into feature/bonus-products-rule-based-v3

sf-shikhar-prasoon committed:
fix header counts

sf-shikhar-prasoon committed:
fix flasing, re-rendering

sf-deepali-bharmal commented on packages/template-retail-react-app/app/utils/bonus-product/business-logic.js:
```diff
@@ -25,6 +25,48 @@ import {
+    const bonusProducts = bonusDiscountLineItem.bonusProducts || []
+    return bonusProducts.length === 0
+}
+
```
Do you also want to check if bonusDiscountLineItem has `promotionId` ?
basically has `promotionId` and does not have `bonusProducts List`

sf-shikhar-prasoon replied:
check added

sf-shikhar-prasoon committed:
Merge branch 'develop' into feature/bonus-products-rule-based-v3

sf-deepali-bharmal commented on packages/template-retail-react-app/app/utils/bonus-product/hooks.js:
```diff
@@ -117,16 +119,55 @@ export const useBasketProductsWithPromotions = (basket) => {
+        ruleBasedPromotions[0] || '',
+        {
+            enabled: ruleBasedPromotions.length > 0,
+            limit: 50
```
How did we land on this limit ?

sf-deepali-bharmal replied:
You can extract it into CONST variable.

sf-shikhar-prasoon replied:
The Api was rejecting a query and in the error, it said limit is 50 . 
I'll add a const.

sf-deepali-bharmal replied:
@sf-shikhar-prasoon `Currently only supports one rule-based promotion at a time.` Why so ?

sf-deepali-bharmal replied:
You are looking for ruleBasedPromotions[0] directly, that can cause runtime error ?

sf-shikhar-prasoon replied:
The ruleBasedPromotions[0] does not throw errors. It could return undefined which is hadled by ruleBasedPromotions[0] || ''

I'm working on adding support for multiple rule-based promotions on the same product...

vmarta replied:
Let's mention in the changelog that we're currently supporting only 1 rule-based promotion at a time.

sf-deepali-bharmal commented on packages/template-retail-react-app/app/hooks/use-rule-based-bonus-products.js:
```diff
@@ -0,0 +1,92 @@
+                    Authorization: token ? `Bearer ${token}` : '',
+                    'Content-Type': 'application/json'
+                }
+            })
```
@sf-shikhar-prasoon Did you check for an `useProductSearch` API in `commerce-sdk-react`  ?

We should try and use that.

sf-deepali-bharmal replied:
Also did you verify if API returns only Bonus products or also returns Qualifiers ? And you need to pass any other `refine` param to filter that ?
That was the issue earlier.

sf-shikhar-prasoon replied:
useProductSearch in commerce-sdk did not allow for pmid or refine=pmid. It was removing anything extra added. So had to switch to using a URL

Did not see any qualifiers. Can you please explain what was the issue last time?

sf-deepali-bharmal replied:
> useProductSearch in commerce-sdk did not allow for pmid or refine=pmid. It was removing anything extra added. So had to switch to using a URL

Is `commerce-sdk` removing `pmid` from `refine`? Then we should update that logic. Making direct http calls is not a good practice.

sf-deepali-bharmal commented on packages/template-retail-react-app/app/hooks/use-rule-based-bonus-products.js:
```diff
@@ -0,0 +1,92 @@
+        queryKey: ['rule-based-bonus-products', promotionId, limit, offset],
+        queryFn: async () => {
+            // Get auth token
+            const token = await getTokenWhenReady()
```
Also I don think we should manually handle tokens. If we use Commerce SDK we don't have to get into handling tokens etc.

sf-shikhar-prasoon replied:
The commerce sdk was not allowing to use refine=pmid=<...> 
It was filtering it out.

sf-shikhar-prasoon replied:
Let me try again to use commerce-sdk...

adamraya replied:
I agree, we should use the `commerce-sdk-react` package instead of making a direct requests to SCAPI. 
The `commerce-sdk-react` is part of the monorepo and if something is wrong, we can fix it directly in the package.

But, something like this should work using `commerce-sdk-react`:

```
import {useProductSearch} from '@salesforce/commerce-sdk-react'
...
const {data, isLoading, error} = useProductSearch(
  {
    parameters: {
      refine: [`pmid=${promotionId}`],
      limit,
      offset
    }
  },
  {enabled: enabled && Boolean(promotionId)}
)
```

sf-shikhar-prasoon replied:
We have changed it to use commerce SDK now

sf-shikhar-prasoon committed:
pr comments + lint

sf-shikhar-prasoon committed:
@W-19980361 fix 2 bugs- selection count & flashing due to re-renders (#3421)

* fix header counts

* fix flasing, re-rendering

* pr comments + lint

sf-shikhar-prasoon committed:
add param &refine=pmpt=bonus for promotion type

sf-deepali-bharmal commented on packages/template-retail-react-app/app/utils/bonus-product/calculations.js:
```diff
@@ -43,13 +43,17 @@ export const getBonusProductCountsForPromotion = (basket, promotionId) => {
+    const selectedBonusItems = bonusProductItems.reduce(
+        (sum, item) => sum + (item.quantity || 0),
+        0
+    )
```
Whats the difference in two ? Why are we changing this?

sf-shikhar-prasoon replied:
bonusProductItems was being used in debugging. 
 Removed this change

sf-deepali-bharmal commented on packages/template-retail-react-app/app/hooks/use-bonus-product-selection-modal/components/bonus-product-item.js:
```diff
@@ -77,7 +93,7 @@ const BonusProductItem = ({
         })
         return defaultSmallImage
-    }, [productData, product])
+    }, [productData, product, productName])
```
This reloads the function if any of these variables are changed right ?
Why do we need to add productName ? wont `productData` already include that ?
May be you can ask cursor.

sf-shikhar-prasoon replied:
Yes, correct. Removed `productName`

sf-deepali-bharmal commented on packages/template-retail-react-app/app/hooks/use-bonus-product-selection-modal/use-bonus-product-data.js:
```diff
@@ -36,15 +38,55 @@ export const useBonusProductData = (modalData) => {
+            enabled: ruleBasedPromotions.length > 0,
+            limit: 50
+        }
+    )
```
Why are we not creating a list of rule based promotionIds ?
And call search API for all promotionIds instead of one at a time ?

sf-deepali-bharmal replied:
According to @sf-shikhar-prasoon  This PR only supports 1 rule based promotion per product. Mentioned the same in changeLog

sf-deepali-bharmal commented on packages/template-retail-react-app/app/components/bonus-product-view-modal/index.jsx:
```diff
@@ -65,6 +61,26 @@ const BonusProductViewModal = ({
+        }),
+        [productViewModalData.product, productViewModalData.isFetching]
+    )
+
```
flashing during fetches must be general problem for both List and Rule based promotions right ?
Why are we adding this now ?

sf-shikhar-prasoon replied:
I think it was there previously. But testing it with 50 selectable products this time brought forward the flashing problem as there were 50 react queries on size selection which caused 50 renders

adamraya commented on packages/template-retail-react-app/app/hooks/use-rule-based-bonus-products.js:
```diff
@@ -0,0 +1,91 @@
+                currency,
+                refine: `pmid=${promotionId}`,
+                limit: String(limit || 25),
+                offset: String(offset || 0)
```
nit: Do we want to reuse the same constants we use for the default search params, or create new ones so values like 25 and 0 aren’t hardcoded and lost in the code?

https://github.com/SalesforceCommerceCloud/pwa-kit/blob/60d9a47ad71824709be52bbe1110b092f7698244/packages/template-retail-react-app/app/constants.js#L18

adamraya commented on packages/template-retail-react-app/app/hooks/use-rule-based-bonus-products.js:
I see we other related bonus product hooks like `useAvailableBonusItemsForProduct`, `useRemainingAvailableBonusProductsForProduct`, etc. in:

https://github.com/SalesforceCommerceCloud/pwa-kit/blob/cca5519d9addda996121dfe9cfd0f45c089dc3a1/packages/template-retail-react-app/app/utils/bonus-product/hooks.js#L120 

But this new `useRuleBasedBonusProducts` hook is separated from rest of bonus product hooks. Should we move it to `/app/utils/bonus-product/hooks.js` for consistency?

sf-shikhar-prasoon replied:
I think it's better to keep it separate. The other file is already 260 files and this one is ~100. Files grow in size over time and we've had to break up files that were over 400-450 lines

adamraya commented on packages/template-retail-react-app/app/utils/bonus-product/hooks.js:
```diff
@@ -117,16 +119,55 @@ export const useBasketProductsWithPromotions = (basket) => {
+        return {
+            [ruleBasedPromotions[0]]: ruleBasedProducts
+        }
+    }, [ruleBasedProducts, ruleBasedPromotions])
```
These logic is duplicated inside the `useRemainingAvailableBonusProductsForProduct` hook:

https://github.com/SalesforceCommerceCloud/pwa-kit/blob/cca5519d9addda996121dfe9cfd0f45c089dc3a1/packages/template-retail-react-app/app/utils/bonus-product/hooks.js#L186-L214

and in `useBonusProductData` hook:
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/cca5519d9addda996121dfe9cfd0f45c089dc3a1/packages/template-retail-react-app/app/hooks/use-bonus-product-selection-modal/use-bonus-product-data.js#L41-L58

Would you consider DRY the code by creating a reusable?

sf-shikhar-prasoon replied:
done!

sf-shikhar-prasoon committed:
pr comments

sf-shikhar-prasoon committed:
use commerce sdk

vmarta commented:
Is this a bug? This video shows how it's possible a product that does _not_ have bonus products could still be associated with bonus products.

https://github.com/user-attachments/assets/d384d0dc-fc10-4154-a97d-5cf116c2b3f8

<img width="716" height="797" alt="Arc 2025-10-23 at 19 14 03" src="https://github.com/user-attachments/assets/4a901b67-55e4-4b4f-9d17-aa8bdcd5805e" />

vmarta commented:
Now that we have modal for choosing the bonus products, let's make sure that its UX is the same/similar to the other modal for editing the cart item.

For bonus products, the content shifts around each time we select a different size, for example. Compare that with how the UX feels like for editing the cart item. Yes, selecting a different size triggers a fetch, but is there a way to keep the content feels more stable?

https://github.com/user-attachments/assets/0ea0b8c2-c319-461f-a992-f02819a01c63

sf-shikhar-prasoon committed:
Merge feature/bonus-products-rule-based-v3 into t/cc-sharks/W-19980361/ruleBasedBugs/main

Resolved conflict in use-rule-based-bonus-products.js by keeping Commerce SDK implementation (useProductSearch) which includes pmpt=bonus parameter.

Merged changes:
- Hybrid proxy support for local development
- Order details multiship and BOPIS support
- Bonus product calculations improvements
- Various CHANGELOG updates

sf-shikhar-prasoon committed:
add const- limit=50  |  replace harded Search Param values

sf-shikhar-prasoon committed:
Merge pull request #3424 from SalesforceCommerceCloud/t/cc-sharks/W-19980361/ruleBasedBugs/main

PR comments for #3418 (PR for merging feature branch bonus-products-rule-based-v3 into develop)

sf-shikhar-prasoon committed:
remove function breakdown change

sf-shikhar-prasoon committed:
update changelog

sf-shikhar-prasoon committed:
refactor. re-use

sf-shikhar-prasoon committed:
Merge branch 'develop' into feature/bonus-products-rule-based-v3

sf-xingquan-jin committed:
fix: prevent promotion message flashing when changing product variant… (#3428)

sf-shikhar-prasoon committed:
fix tests

sf-shikhar-prasoon committed:
Merge branch 'develop' into feature/bonus-products-rule-based-v3

sf-cboscenco committed:
Rule-Based Bonus Product Promotions Enhancement

This PR enhances the existing rule-based bonus product promotions feature by adding proper variant eligibility checking. Previously, rule-based promotions could not accurately determine which specific product variants qualified for a promotion. This enhancement introduces a qualifying products lookup system that uses the product search API to fetch and validate eligible variants, ensuring that only products that truly qualify can trigger bonus product selection UI and functionality.

Key Changes:

1. New Hook: useRuleBasedQualifyingProducts
   - Added to app/hooks/use-rule-based-bonus-products.js
   - Fetches qualifying products for rule-based promotions using productSearch API
   - Uses refinement parameters: pmid=${promotionId} and pmpt=qualifying
   - Returns a Set of qualifying product IDs for efficient lookup
   - Supports variant eligibility checking by examining both variant and master product IDs

2. Enhanced Promotion Eligibility Logic
   - Updated getPromotionIdsForProduct() in app/utils/bonus-product/common.js
   - Distinguishes between list-based and rule-based promotions
   - Added fallback logic to check master product IDs when variants don't directly qualify
   - Filters promotion IDs to only return those the product actually qualifies for

3. Qualifying Products Map Integration
   - Added ruleBasedQualifyingProductsMap parameter throughout the bonus product utility chain
   - Updated useBasketProductsWithPromotions() hook to fetch and return qualifying products map
   - Propagated the map through all relevant functions and components

4. Updated Function Signatures
   - isAutomaticPromotion()
   - shouldShowBonusProductSelection()
   - getBonusProductsForSpecificCartItem()
   - getBonusProductsInCartForProduct()
   - getAvailableBonusItemsForProduct()
   - getRemainingAvailableBonusProductsForProduct()

5. Component Updates
   - Cart Page: Destructures and passes ruleBasedQualifyingProductsMap
   - Add to Cart Modal: Uses qualifying products map for eligibility checks
   - Cart Product List: Accepts and forwards ruleBasedQualifyingProductsMap prop

6. Comprehensive Test Updates
   - Updated all test files to distinguish list-based vs rule-based promotions
   - Added new test cases for rule-based promotion eligibility scenarios
   - Enhanced test coverage across all bonus product utility test files

Backward Compatibility:
- All changes are backward compatible with existing list-based promotions
- Default parameter values ensure functions work without the new map
- Existing tests updated to explicitly mark promotions as list-based

sf-cboscenco committed:
Merge latest from the feature branch

sf-xingquan-jin committed:
fix: prevent promotion message flashing when changing product variant (#3431)

* fix: prevent promotion flash by keeping previous data during variant changes

- Add keepPreviousData option parameter to useProductViewModal hook
- Pass keepPreviousData: true in BonusProductViewModal to maintain UI stability
- Prevents promotion message from clearing while fetching new variant data

* chore: trigger CI

* remove those stable references

* revert last changes

sf-cboscenco committed:
Merge pull request #3432 from SalesforceCommerceCloud/t/cc-sharks/W-20037720

Rule-Based Bonus Product Promotions Enhancement

sf-shikhar-prasoon committed:
bundle size limit update

ddiazccrz committed:
@W-20037720: Not allow duplicate items to appear in the cart (#3433)

* Fixed duplicating bonus products in the UI

* Fixed issue with list based not rendering properly

* Update cart-product-list-with-grouped-bonus-products.jsx

Signed-off-by: Daniel Diaz <38542490+ddiazccrz@users.noreply.github.com>

* Fixed lint errors

---------

Signed-off-by: Daniel Diaz <38542490+ddiazccrz@users.noreply.github.com>

sf-deepali-bharmal commented on packages/template-retail-react-app/app/components/bonus-product-view-modal/index.jsx:
```diff
@@ -110,31 +129,24 @@ const BonusProductViewModal = ({
-                getRemainingAvailableBonusProductsForProduct
-            ),
-        [basket, product]
+        [intl, stableProductViewModalData?.product?.name, formatMessage]
```
`intl ?` What does this do ?

ddiazccrz replied:
intl changes when the users language changes to recalculate the formatted message, could just use [formatMessage, stableProductViewModalData?.product?.name]

ddiazccrz reviewed (approved):
I am approving this with the understanding that there are a few minor existing bugs (there with bonus products before this rule based bonus products) that will be fixed after. @sf-shikhar-prasoon lets make sure we make work items to address the lower priority issues.

sf-deepali-bharmal commented on packages/template-retail-react-app/app/utils/bonus-product/calculations.js:
```diff
@@ -43,6 +43,7 @@ export const getBonusProductCountsForPromotion = (basket, promotionId) => {
 
     // Count selected items for this promotion (all bonus items with this promotion's bonusDiscountLineItemIds)
     const promotionBonusLineItemIds = promotionBonusItems.map((item) => item.id).filter(Boolean)
+
```
This file can be reverted.

sf-deepali-bharmal reviewed (approved):
I am approving the PR but please test this feature branch thoroughly, preferably maintain a doc with tested scenario's.
Mix of rule based, list based, multiple rules or list based promo's in cart etc end to end.
